src/include/migraphx/op/reshape.hpp

```diff
@@ -64,45 +65,83 @@ struct reshape
 
     std::string name() const { return "reshape"; }
 
-    shape dyn_compute_shape(shape s0) const
+    // Assumes that the shape from the `dims` attribute will be valid at run-time.
+    // Makes no checks for the validity of the `dims` attribute for the given input shape.
+    shape dyn_1arg_compute_shape(shape s0) const
     {
-        auto dyn_dims = s0.dyn_dims();
-        // track number of fixed elements in input and output
-        std::size_t num_dims_ele = 1;
-        std::size_t num_dd_ele   = 1;
-        for(std::size_t i = 0; i < dyn_dims.size(); ++i)
+        auto input_dyn_dims = s0.dyn_dims();
+        bool has_negative_dim_attr = false;
+        // construct output dynamic shape from dims attribute
+        std::vector<shape::dynamic_dimension> output_dyn_dims(dims.size());
+        for(int i = 0; i < dims.size(); ++i)
         {
-            if(dyn_dims[i].is_fixed())
+            int64_t d = dims.at(i);
+            if(d == 0)
+            {
+                output_dyn_dims.at(i) = input_dyn_dims.at(i);
+            }
+            else if(d == -1)
             {
-                num_dims_ele *= dims[i];
-                num_dd_ele *= dyn_dims[i].min;
+                has_negative_dim_attr = true;
+                output_dyn_dims.at(i) = {1, 1};
             }
             else
             {
-                if(dims[i] != 0 and dims[i] != -1)
-                {
-                    MIGRAPHX_THROW(
-                        "Reshape: Non-fixed dynamic_dimension doesn't match with 0 or -1 "
-                        "output dimension");
-                }
+                std::size_t u_dim     = d;
+                output_dyn_dims.at(i) = {u_dim, u_dim};
             }
         }
-        if(num_dims_ele != num_dd_ele)
+
+        if(has_negative_dim_attr)
         {
-            MIGRAPHX_THROW("Reshape: Number of fixed elements must match. Input: " +
-                           std::to_string(num_dd_ele) + " Output: " + std::to_string(num_dims_ele));
+            // comparing the -1 dimension against the other dimensions
+            auto neg_dim_num = std::distance(this->dims.begin(),
+                                             std::find(this->dims.begin(), this->dims.end(), -1));
+
+            // unsigned int wraparound check, false = no wraparound
+            auto uint_wraparound_check = [](std::size_t a, std::size_t b) {
+                std::size_t c = a * b;
+                return a != 0 and c / a != b;
+            };
+
+            // accumulate the minimum and maximum elements in the dimensions before the -1 dimension
+            std::size_t min_cur_elements = 1;
+            std::size_t max_cur_elements = 1;
+            std::size_t max_int          = std::numeric_limits<std::size_t>::max();
+            for(const auto& dd : output_dyn_dims)
+            {
+                min_cur_elements = uint_wraparound_check(min_cur_elements, dd.min)
```

Comment:
What you want is [`std::mul_sat`](https://en.cppreference.com/w/cpp/numeric/mul_sat) which unfortunately we cant use that yet, but we can add our own implementation:

```cpp
template<class T>
constexpr T mul_sat(T a, T b) noexcept
{
    T c;
    if (not __builtin_mul_overflow(a, b, &c))
        return c;
    if constexpr(std::is_unsigned<T>{})
        return std::numeric_limits<T>::max();
    if (a < 0 xor b < 0)
        return std::numeric_limits<T>::min();
    return std::numeric_limits<T>::max();
}
```

The builtin should be available in both gcc and clang. Maybe add this under a header such as `sat_ops.hpp`.

Reply:
What about Windows?

Reply:
> What about Windows?

We compile with clang on windows so the intrinsic shouldn't be a problem.